internal/stacks/stackmigrate/resources.go

```diff
@@ -0,0 +1,373 @@
+		if schema == nil {
+			m.emitDiags(diags.Append(&hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Resource type not found",
+				Detail:   fmt.Sprintf("Resource type %s not found in provider schema.", resource.Addr.Resource.Type),
+				Subject:  target.ModuleConfig.SourceAddrRange.Ptr(),
+			}))
+			continue
+		}
+
+		for instanceKey, instance := range resource.Instances {
+			instanceAddr := stackaddrs.AbsResourceInstance{
+				Component: component,
+				Item:      componentAddr.Instance(instanceKey),
+			}
+
+			m.emit(&stackstate.AppliedChangeResourceInstanceObject{
+				ResourceInstanceObjectAddr: stackaddrs.AbsResourceInstanceObject{
+					Component: instanceAddr.Component,
+					Item:      instanceAddr.Item.DeposedObject(addrs.NotDeposed),
+				},
+				NewStateSrc:        instance.Current,
+				ProviderConfigAddr: providerAddr,
+				Schema:             schema,
+			})
+
+			for deposedKey, deposed := range instance.Deposed {
+				m.emit(&stackstate.AppliedChangeResourceInstanceObject{
+					ResourceInstanceObjectAddr: stackaddrs.AbsResourceInstanceObject{
+						Component: instanceAddr.Component,
+						Item:      instanceAddr.Item.DeposedObject(deposedKey),
+					},
+					NewStateSrc:        deposed,
+					ProviderConfigAddr: providerAddr,
+					Schema:             schema,
+				})
+			}
+		}
+	}
+	return components
+}
+
+// search searches for the state resource in the resource mappings and when found, converts and returns the relevant
+// stackResource.
+//
+// If the resource or module is nested within the root module, they will be migrated to the component with the address structure retained.
+// For example, a resource with the address module.my_module.module.child.aws_instance.foo will be migrated to
+// component.my_component.module.child.aws_instance.foo if the corresponding map key is found.
+// E.g module.child.aws_instance.foo will be replaced with component.child.aws_instance.foo
+func (m *migration) search(resource addrs.AbsResource, resources map[string]string, modules map[string]string) (*stackResource, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	ret := &stackResource{
+		ContainingModule: resource.Module.Module(),
+	}
+
+	parseComponentInstance := func(target string) (Instance, tfdiags.Diagnostics) {
+		fullTarget := "component." + strings.TrimPrefix(target, "component.")
+		// if strings.Contains(target, "stack.") {
+		// 	fullTarget = target
+		// }
+		if len(strings.Split(fullTarget, ".")) > 2 {
+			diags = diags.Append(tfdiags.Sourceless(tfdiags.Error, "Invalid component instance", fmt.Sprintf("Only root component instances are allowed, got %q", target)))
+			return Instance{}, diags
+		}
+		inst, _, diags := stackaddrs.ParseAbsComponentInstanceStrOnly(fullTarget)
+		return inst, diags
+	}
+
+	if resource.Module.IsRoot() {
+		// If there is no resource mapping, we check for a root module mapping
```

Comment:
TODO: Need to confirm if the consuming team wants this

Reply:
I think we should be explicit here to begin with - it is safer to loosen restrictions later than to tighten them so I think either return an error for now and we can add this later if needed.